menpo/model/linear.py

```diff
@@ -208,6 +208,49 @@ def reconstruct_vectors(self, vectors):
         """
         return self.instance_vectors(self.project_vectors(vectors))
 
+    def impute(self, vector, mask):
+        """
+        Projects the `vector` onto a subset of the linear space and
+        rebuilds from the weights found.
+
+        Parameters
+        ----------
+        vector : ``(n_available_features,)`` `ndarray`
+            A vector to impute. This is a subset of the original
+            feature space, which is specified by the provided mask.
+        mask : ``(n_features,)`` `bool ndarray`
+            A boolean mask that is True where the feature exists in the input
+            vectors.
+
+        Returns
+        -------
+        imputed : ``(n_features,)`` `ndarray`
+            The imputed vector.
+        """
+        return self.impute_vectors(vector[None, :], mask).ravel()
+
+    def impute_vectors(self, vectors, mask):
+        """
+        Projects the `vectors` onto a subset of the linear space and
+        rebuilds vectors from the weights found.
+
+        Parameters
+        ----------
+        vectors : ``(n_vectors, n_available_features)`` `ndarray`
+            A set of vectors to impute. This is a subset of the original
+            feature space, which is specified by the provided mask.
+        mask : ``(n_features,)`` `bool ndarray`
+            A boolean mask that is True where the feature exists in the input
+            vectors.
+
+        Returns
+        -------
+        imputed : ``(n_vectors, n_features)`` `ndarray`
+            The imputed vectors.
+        """
+        imputed_weights = np.dot(vectors, self.components[:, mask].T)
```

Comment:
@patricksnape perhaps I'm missing something (not familiar with the definition of impute in this context!) but I think this isn't quite the optimal solution. I had to implement this for the 3DMM stuff here and originally _exactly_ did this, before getting schooled by wiser (Greek) minds that this ain't optimal. https://github.com/menpo/menpo3d/blob/master/menpo3d/morphablemodel/shapemodel.py#L58
(Ignore the fact that it's shape specific!)

Key point is for the subset of the basis left after masking (here `U`) the basis left is not orthogonal, so `U @ U.T` is not identity. Due to this, we have to actually solve the linear problem `v = U @ w + m`, we can't just do the dot product.

In practice, for my problems I actually found the dot worked pretty well, but technically the `lstsq` solve is what we should be doing...thoughts?